grug/types/src/events/flattened.rs

```diff
@@ -0,0 +1,892 @@
+
+    flat_events
+}
+
+// -------------------------- impl Flat for Status --------------------------
+
+impl<T> FlattenStatus for EventStatus<T>
+where
+    T: Flatten,
+{
+    fn flatten_status(
+        self,
+        parent_id: &EventId,
+        next_id: &mut EventId,
+        commitment: FlatCommitmentStatus,
+    ) -> Vec<FlatEventInfo> {
+        match self {
+            EventStatus::Ok(event) => {
+                event.flatten(parent_id, next_id, commitment, FlatEventStatus::Ok)
+            },
+            EventStatus::Failed { event, error } => event.flatten(
+                parent_id,
+                next_id,
+                commitment,
+                FlatEventStatus::Failed(error),
+            ),
+            EventStatus::NestedFailed(event) => event.flatten(
+                parent_id,
+                next_id,
+                commitment,
+                FlatEventStatus::NestedFailed,
+            ),
+            EventStatus::NotReached => vec![],
+        }
+    }
+}
+
+impl FlattenStatus for SubEventStatus {
+    fn flatten_status(
+        self,
+        parent_id: &EventId,
+        next_id: &mut EventId,
+        commitment: FlatCommitmentStatus,
+    ) -> Vec<FlatEventInfo> {
+        let (event, commitment, status) = match self {
+            SubEventStatus::Ok(event) => (event, commitment, FlatEventStatus::Ok),
+            SubEventStatus::NestedFailed(event) => {
+                (event, commitment, FlatEventStatus::NestedFailed)
+            },
+            SubEventStatus::Failed { event, error } => {
+                (event, commitment, FlatEventStatus::Failed(error))
+            },
+            // SubEventStatus::Handled is a particular case.
+            // It means that a submsg fails but the error has been handled on reply.
+            // In this case, the commitment status is Failed regardless of the original commitment status.
+            SubEventStatus::Handled { event, error } => (
+                event,
+                FlatCommitmentStatus::Failed,
+                FlatEventStatus::Handled(error),
+            ),
+        };
+
+        event.flatten(parent_id, next_id, commitment, status)
+    }
+}
+
+impl FlattenStatus for MsgsAndBackrunEvents {
+    fn flatten_status(
+        self,
+        parent_id: &EventId,
+        next_id: &mut EventId,
+        commitment: FlatCommitmentStatus,
+    ) -> Vec<FlatEventInfo> {
+        let mut events = vec![];
+
+        for (msg_idx, msgs) in self.msgs.into_iter().enumerate() {
+            next_id.message_index = Some(msg_idx as u32);
+
+            let i_events = msgs.flatten_status(parent_id, next_id, commitment.clone());
+
+            next_id.increment_idx(&i_events);
+            events.extend(i_events);
+        }
+        next_id.message_index = None;
+
+        events.extend(self.backrun.flatten_status(parent_id, next_id, commitment));
+
+        events
+    }
+}
+
+// -------------------------- impl Flat for Events --------------------------
```

Comment:
Can you extract the trait to a separate file? Put the types in `types/src/events/flat.rs` and the trait in `types/src/events/flatten.rs`